Лаб. 2 6413 Волков Владислав

comandir26 opened the pull request:
**ФИО:** Волков Владислав  
**Номер группы:** 6413  
**Номер лабораторной:** 2  
**Номер варианта:** 63 
**Краткое описание предметной области:** Пункт велопроката
**Краткое описание добавленных фич:** Реализованы репозитории для всех доменных сущностей, инициализирующиеся тестовыми данным из InMemorySeeder, были созданы DTO классы, реализованы сервисы, предоставляющие CRUD операции для работы со всеми сущностями, реализованы контроллеры для всех сущностей с полным набором HTTP методов, также полностью переработаны тесты.

comandir26 committed:
Initial commit

comandir26 committed:
Add classes

comandir26 committed:
Fix classes, add comments

comandir26 committed:
Add test data

comandir26 committed:
Add tests

comandir26 committed:
Fix tests and add comments

comandir26 committed:
Add dotnet_tests.yml for automatic testing

comandir26 committed:
Fix dotnet_tests

comandir26 committed:
Fix dotnet_tests again

comandir26 committed:
Move github directory

comandir26 committed:
Move github directory and change dotnet_tests file again x2

comandir26 committed:
Small fixes

comandir26 committed:
Corrected all comments

comandir26 committed:
Updated readme and comments

comandir26 committed:
Marked methods as static where needed

comandir26 committed:
Add IRepository interface class for inmemory repo classes, also add Bikes.Infrastructure.InMemory project for repo classes

comandir26 committed:
Implement inmemory repo classess

comandir26 committed:
Add DTO classes, add interfaces for service classes and their implementation

comandir26 committed:
Add Bikes.Api.Host project and comments

comandir26 committed:
Update tests and fix comments

comandir26 committed:
Update readme

DmitryKrakhmalev commented on Bikes/Bikes.Contracts/Dto/BikeCreateUpdateDto.cs:
DTO лучше вынести в отдельный проект из Allpication. К примеру Bikes.Contracts.
Зачем это делается - в дальнейшем если у вас появится необходимость оперировать этими DTO вы сможете ссылаться на данный проект, в котором не будет лишней логики по сервисам и прочему.

DmitryKrakhmalev commented on Bikes/Bikes.Application/Services/AnalyticsService.cs:
```diff
@@ -0,0 +1,137 @@
+    /// <param name="bikeModelRepository"></param>
+    /// <param name="rentRepository"></param>
+    /// <param name="renterRepository"></param>
+    public AnalyticsService(
```
Используйте primary constructor тут и в других файлах сервисов. От списков репозиториев можно будет избавиться и использовать объявленные в конструкторе.

DmitryKrakhmalev commented on Bikes/Bikes.Application/Services/AnalyticsService.cs:
```diff
@@ -0,0 +1,137 @@
+    {
+        return _bikeRepository.ReadAll()
+            .Where(bike => bike.Model.Type == BikeType.Sport)
+            .ToList();
```
Упростить инициализацию коллекций. Тут и далее по коду.

```suggestion
        return [.. bikeRepository.ReadAll().Where(bike => bike.Model.Type == BikeType.Sport)];
```

DmitryKrakhmalev commented on Bikes/Bikes.Application/Services/AnalyticsService.cs:
```diff
@@ -0,0 +1,137 @@
+                  x => x.ModelId,
+                  model => model.Id,
+                  (x, model) => model)
+            .ToList();
```
Аналогично. Также упростить коллекции в других местах где это требуется.

DmitryKrakhmalev commented on Bikes/Bikes.Application/Services/BikeService.cs:
```diff
@@ -0,0 +1,98 @@
+    {
+        var model = _bikeModelRepository.Read(bikeDto.ModelId);
+        if (model == null)
+            throw new ArgumentException($"BikeModel with id {bikeDto.ModelId} not found");
```
Можно упростить:

```suggestion
        var model = bikeModelRepository.Read(bikeDto.ModelId)
            ?? throw new ArgumentException($"BikeModel with id {bikeDto.ModelId} not found");
```

DmitryKrakhmalev commented on Bikes/Bikes.Application/Services/RentService.cs:
```diff
@@ -0,0 +1,112 @@
+    {
+        var bike = _bikeRepository.Read(rentDto.BikeId);
+        if (bike == null)
+            throw new ArgumentException($"Bike with id {rentDto.BikeId} not found");
```
Аналогично. И по коду в других местах.

DmitryKrakhmalev commented on Bikes/Bikes.Infrastructure.InMemory/Repositories/InMemoryBikeModelRepository.cs:
```diff
@@ -0,0 +1,91 @@
+/// </summary>
+public class InMemoryBikeModelRepository : IRepository<BikeModel, int>
+{
+    private readonly List<BikeModel> _items = [];
```
Можно не инициализировать, он у вас в конструкторе инициализируется.

```suggestion
    private readonly List<BikeModel> _items;
```

DmitryKrakhmalev commented on Bikes/Bikes.Api.Host/Controllers/AnalyticsController.cs:
```diff
@@ -0,0 +1,140 @@
+    /// A method that returns information about all sports bikes
+    /// </summary>
+    [HttpGet("sport-bikes")]
+    public IActionResult GetSportBikes()
```
Вижу что в ваших методах контроллеров имеется обработка ошибок и логирование - это хорошо.
Также вижу что у вас имеется валидация входных данных.

Что необходимо доработать:
1. Снабдить атрибутами `ProducesResponseType` с указанием возможных статусов ответа;
2. Вместо `IActionResult` возвращаем типизированный `ActionResult<T>`, где `T` - это DTO класс а не класс модели.

DmitryKrakhmalev commented on Bikes/Bikes.Api.Host/Controllers/AnalyticsController.cs:
```diff
@@ -0,0 +1,140 @@
+        try
+        {
+            logger.LogInformation("Getting rental duration statistics");
+            var stats = service.GetRentalDurationStats();
```
Можно сразу разобрать. В целом было бы не плохо если для этого был выделен отдельный DTO класс:

```suggestion
            var (min, max, avg) = service.GetRentalDurationStats();
```

DmitryKrakhmalev commented on Bikes/Bikes.Api.Host/Controllers/AnalyticsController.cs:
```diff
@@ -0,0 +1,140 @@
+                min = stats.min,
+                max = stats.max,
+                average = stats.avg
+            });
```
Заведите DTO. Не нужно оперировать анонимными типами. Возвращайте ActionResult с заданным типом.

DmitryKrakhmalev reviewed (changes requested):
Необходимо устранить замечания.

comandir26 committed:
add primary constructors, collection expressoins
use ProducesResponseTypeautomapper, automapper
all services now returning DTO
and small fixes

comandir26 commented:
Исправил все, что требовалось
1) Вынес DTO в отдельный проект Bikes.Contracts и завел RentalDurationStatsDto
2) Заменил обычные конструкторы на primary
3) Упростил инициализации коллекций, где это возможно 
4) Добавил в методы контроллеров атрибуты ProducesResponseType для статусов ответа, также заменил анонимные объекты на ProblemDetails
5) Заменил IActionResult на ActionResult<T> с конкретными типами возвращаемых DTO
6) Также сервисы теперь возвращают DTO, а не доменные модели
7) В сервисах использую AutoMapper, где это возможно, но в AnalyticsService использую ручной маппинг для упрощения unit-тестирования

DmitryKrakhmalev commented on Bikes/Bikes.Infrastructure.InMemory/Repositories/InMemoryBikeModelRepository.cs:
```diff
@@ -0,0 +1,91 @@
+    /// <returns>List of existing objects</returns>
+    public List<BikeModel> ReadAll()
+    {
+        return _items;
```
Не очень безопасно отдавать саму коллекцию. А если кто то в обход методов репозитория решит этот список обрабатывать. Лучше отдавайте копию. Аналогично поправить в других репозиториях.
```suggestion
        return [.. _items];
```

DmitryKrakhmalev commented on Bikes/Bikes.Contracts/Dto/BikeDto.cs:
```diff
@@ -0,0 +1,22 @@
+/// <summary>
+/// DTO class for the Bike class
+/// </summary>
+public class BikeDto
```
У вас для каждого класса заведено по одной DTO, при этом получается не возможным получить идентификатор объекта. Каким образом планируется передача информации об идентификаторе объекта.
Обычно рекомендуется заводить по две DTO (на создание/изменение и на получение данных)

DmitryKrakhmalev commented on Bikes/Bikes.Application/Services/AnalyticsService.cs:
```diff
@@ -0,0 +1,149 @@
+                      BrakeType = model.BrakeType,
+                      Year = model.Year,
+                      RentPrice = model.RentPrice
+                  })];
```
Не очень продуктивно получилось. У вас имеется информация по моделям но вы ее отбрасываете оставляя только идентификатор. Затем вы опять подключаете эти данные.
Я бы вам посоветовал вообще убрать список `models` а группировать по всему классу `rent.Bike.Model`.
Тогда у вас Join будет не нужен. Просто через Select приведите к DTO.

DmitryKrakhmalev commented on Bikes/Bikes.Application/Services/AnalyticsService.cs:
```diff
@@ -0,0 +1,149 @@
+            })
+            .OrderByDescending(x => x.TotalProfit)
+            .Take(5)
+            .Join(models,
```
Попробовать переписать по аналогии с замечанием выше.

DmitryKrakhmalev commented on Bikes/Bikes.Application/Interfaces/IAnalyticsService.cs:
Я бы предложил интерфейсы погрузить в отдельную папку Intrerfaces чтобы не было мешанины интерфейсов от реализаций.

DmitryKrakhmalev commented on Bikes/Bikes.Application/Services/BikeModelService.cs:
```diff
@@ -0,0 +1,82 @@
+        {
+            var dto = mapper.Map<BikeModelDto>(model);
+            return dto;
+        }).ToList();
```
Упростить инициализацию коллекций

DmitryKrakhmalev commented on Bikes/Bikes.Api.Host/Extensions/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,34 @@
+    {
+        services.AddAutoMapper(AppDomain.CurrentDomain.GetAssemblies());
+
+        services.AddSingleton<IRepository<Bikes.Domain.Models.Bike, int>, InMemoryBikeRepository>();
```
```suggestion
        services.AddSingleton<IRepository<Domain.Models.Bike, int>, InMemoryBikeRepository>();
```
Ниже аналогично.

DmitryKrakhmalev reviewed (changes requested):
Необходимо устранить замечания. Также необходимо протестировать Api. У меня при попытке получения BikModels (да и не только) постоянно возникала 500 ошибка.
<img width="1395" height="564" alt="image" src="https://github.com/user-attachments/assets/1abb02e6-e264-47a7-aa03-98139113f221" />

comandir26 committed:
Fix api

comandir26 committed:
Fix dto's and small fixes

comandir26 committed:
Fix tests

comandir26 commented:
1) Протестировал Api, сейчас все работает, были проблемы с версиями пакетов для Automapper и отсутствовали Profile классы
2) Завел по 2 DTO
3) Остальные замечания исправил

DmitryKrakhmalev commented on Bikes/Bikes.Contracts/Dto/RentGetDto.cs:
```diff
@@ -0,0 +1,34 @@
+﻿using Bikes.Domain.Models;
```
Неиспользуемый юзинг.

DmitryKrakhmalev commented on Bikes/Bikes.Application/Services/AnalyticsService.cs:
```diff
@@ -0,0 +1,125 @@
+/// </summary>
+public class AnalyticsService(
+    IRepository<Bike, int> bikeRepository,
+    IRepository<BikeModel, int> bikeModelRepository,
```
Не используется, нужно убрать.

DmitryKrakhmalev commented on Bikes/Bikes.Contracts/Dto/BikeModelCreateUpdateDto.cs:
```diff
@@ -0,0 +1,44 @@
+    /// <summary>
+    /// Model's production year
+    /// </summary>
+    public required string Year { get; set; }
```
Не очень понимаю зачем год выпуска вы решили хранить как строку. Может лучше переделать на int?

DmitryKrakhmalev commented on Bikes/Bikes.Api.Host/Controllers/BikeModelsController.cs:
```diff
@@ -0,0 +1,204 @@
+    [ProducesResponseType(typeof(CreatedAtActionResult), StatusCodes.Status201Created)]
+    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status400BadRequest)]
+    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status500InternalServerError)]
+    public ActionResult<CreatedAtActionResult> CreateBikeModel([FromBody] BikeModelCreateUpdateDto bikeModelDto)
```
Вам необходимо указать конкретный тип возвращаемого результата (в данном случае `BikeModelGetDto`, так как вызывается метод `GetBikeModel`. Аналогично в ProducesResponseType.
У меня даже сваггеру поплохело и он дико тормозил из за этого. Естественно создание объектов отваливалось с ошибкой.

DmitryKrakhmalev commented on Bikes/Bikes.Api.Host/Controllers/BikeModelsController.cs:
```diff
@@ -0,0 +1,204 @@
+        {
+            logger.LogInformation("Creating new bike model of type {BikeType}", bikeModelDto.Type);
+
+            if (!ModelState.IsValid)
```
По сути тут ничего особо не валидируется так как у вас с DTO нет атрибутов валидации. Очень напрашивается дополнительное ограничения на некоторые элементы (к примеру вес, размер колес, год, ЦЕНУ наконец то).
Что будет если я отрицательную стоимость укажу. Пользуйтесь валидационными атрибутами, такими как Range и другие. 
https://learn.microsoft.com/ru-ru/aspnet/core/mvc/models/validation?view=aspnetcore-8.0#validation-attributes

DmitryKrakhmalev reviewed (changes requested):
Проработайте валидацию входящих DTO, а также устраните замечания.

comandir26 committed:
Validation of DTOs and fix creating objects

comandir26 committed:
some validation checks

comandir26 commented:
1) Добавил валидационные атрибуты для DTO и протестировал их
2) Исправил методы создания в контроллерах, указав конкретные типы возвращаемых значений
3) Остальные замечания устранил

comandir26 committed:
Fix create methods

comandir26 commented:
Немного исправил создание в контроллерах, убрав лишний запрос для получения созданной записи

DmitryKrakhmalev reviewed (approved):
Засчитано
